Add support for CUDA Array Interface for interoperability with other Python libraries (resolves #633)

JRV7903 opened the pull request:
- Added the __cuda_array_interface__ property to the Array class to expose CUDA array data for interoperation.
- Implemented getDevicePointer() in the Array class to retrieve device pointers.
- Updated the cudaBackend.cc and backend.h files to handle the device pointers and support CUDA-specific array operations.
- Modified model_preprocessor.py to enable the inclusion of the __cuda_array_interface__ property for array interoperability.
- Created a new test file test_cuda_interface.py for verifying the changes and ensuring the correct functionality of the CUDA array interface.

neworderofjamie commented on tests/features/test_cuda_interface.py:
```diff
@@ -0,0 +1,64 @@
+
+from pygenn import genn_model
+
+@pytest.mark.parametrize("precision,np_dtype", [("float", np.float32), ("double", np.float64)])
+def test_cuda_array_interface(make_model, backend, precision, np_dtype):
```
there is a mechanism for selecting backends tests run for in https://github.com/genn-team/genn/blob/master/tests/features/conftest.py#L25-L53 which you could extend with a ``backend_cuda`` fixture

neworderofjamie replied:
the seperate numpy dtype is unecessary - your range will be cast to ``precision`` anyway
```suggestion
@pytest.mark.parametrize("precision", ["float", "double"])
def test_cuda_array_interface(make_model, backend, precision):
```

neworderofjamie commented on tests/features/test_cuda_interface.py:
```diff
@@ -0,0 +1,64 @@
+import sys
+
+try:
+    import cupy as cp
```
it would be good to:

1. Add cupy to the list of test dependencies at https://github.com/genn-team/genn/blob/master/setup.py#L316
2. Set the CI to install the dependency at https://github.com/genn-team/genn/blob/master/Jenkinsfile#L196

neworderofjamie commented on src/genn/backends/cuda/backend.cc:
```diff
@@ -55,187 +55,170 @@ const EnvironmentLibrary::Library doubleRandomFunctions = {
           VarLocation location, bool uninitialized)
-    :   ArrayBase(type, count, location, uninitialized), m_DevicePointer(nullptr)
-    {
-        // Allocate if count is specified
```
Can you restore the comments?

neworderofjamie commented on src/genn/backends/cuda/backend.cc:
```diff
@@ -55,187 +55,170 @@ const EnvironmentLibrary::Library doubleRandomFunctions = {
-            if((offset + count) > getCount()) {
-                throw std::runtime_error("Cannot pull slice that overflows array");
-            }
+//! Copy a 1D slice of elements to device 
+/*! \param offset   Offset in elements to start copying from
+    \param count    Number of elements to copy*/
```
these comments get used to generate API documentation, can you move them to the header?

neworderofjamie commented on src/genn/backends/cuda/backend.cc:
```diff
@@ -55,187 +55,170 @@ const EnvironmentLibrary::Library doubleRandomFunctions = {
+    throw std::runtime_error("CUDA arrays have no host objects");
+}
 
+std::byte *Array::getDevicePointer() const
+{
+    return m_DevicePointer;
+}
```
I think this can probably be declared inline in the header

neworderofjamie reviewed (changes requested):
This looks really great - nice work! Just a few minor suggestions

neworderofjamie commented on tests/features/test_cuda_interface.py:
```diff
@@ -0,0 +1,64 @@
+            "TauRefrac": 0.0
+        }, 
+        {
+            "V": -65.0,
```
you can assign the range of initial values directly here:
```suggestion
            "V": np.linspace(-70.0, -60.0, 100),
```

neworderofjamie commented on pygenn/model_preprocessor.py:
```diff
@@ -64,6 +64,33 @@ def pull_from_device(self):
+            model = self.group._model
+            resolved_type = (self.type if isinstance(self.type, ResolvedType)
+                            else self.type.resolve(model._type_context))
+            dtype = model.genn_types[resolved_type]
```
This is the only actual bug I found - with this it works great 😄 
```suggestion
            dtype = np.type(model.genn_types[resolved_type])
```

JRV7903 commented:
Thank you for reviewing the PR. I’ve fixed the bug so that CuPy now correctly receives a NumPy datatype. I’ve also implemented the other requested changes.

neworderofjamie reviewed (changes requested):
Amazing! If you can accept my one suggestion, I will merge
